tensorflow-framework/src/main/java/org/tensorflow/framework/metrics/impl/Reduce.java

```diff
@@ -0,0 +1,239 @@
+    super(tf, name, seed);
+    this.reduction = reduction;
+    this.totalName = this.getVariableName(TOTAL);
+    this.countName = this.getVariableName(COUNT);
+    this.type = type;
+    setupVars();
+  }
+  /** Initializes the Variables */
+  private void setupVars() {
+    if (total == null) {
+      total = getTF().withName(totalName).variable(Shape.scalar(), type);
+    }
+    if (reduction == MetricReduction.SUM_OVER_BATCH_SIZE
+        || reduction == MetricReduction.WEIGHTED_MEAN) {
+      if (count == null) {
+        count = getTF().withName(countName).variable(Shape.scalar(), type);
+      }
+    }
+  }
+
+  /** {@inheritDoc} */
+  public Op resetStates() {
+    List<Op> controls = new ArrayList<>();
+    if (total != null) {
+      controls.add(
+          getTF().assign(total, CastHelper.cast(getTF(), getTF().constant(0), total.type())));
+    }
+    if (count != null) {
+      controls.add(
+          getTF().assign(count, CastHelper.cast(getTF(), getTF().constant(0), count.type())));
+    }
+    return getTF().withControlDependencies(controls).noOp();
+  }
+
+  /**
+   * Updates the metric variables based on the inputs. At least one input arg required for <code>
+   * values</code>, an optional additional input for the <code>sampleWeights</code>
+   *
+   * @param values the inputs to be passed to update state, this may not be null
+   * @param sampleWeights sample weights to be applied to values, may be null.
+   * @return the result with a control dependency on update state Operands
+   * @throws IllegalArgumentException if values is null
+   */
+  @Override
+  public List<Op> updateStateList(Operand<U> values, Operand<T> sampleWeights) {
+
+    if (values == null) {
+      throw new IllegalArgumentException("values is required.");
+    }
+    List<Op> updateOperations = new ArrayList<>();
+    // cast everything to match the variables
+    Operand<U> lSampleWeights = null;
+    Operand<U> lValues = values;
+
+    if (sampleWeights != null) {
+      lSampleWeights = CastHelper.cast(getTF(), sampleWeights, lValues.type());
+      LossTuple<U> tuple =
+          LossesHelper.squeezeOrExpandDimensions(getTF(), null, lValues, lSampleWeights);
+      lValues = tuple.getTarget();
+      lSampleWeights = tuple.getSampleWeights();
+      try {
+
+        Op broadcastWeightsCheck = MetricsHelper.broadcastWeights(getTF(), lSampleWeights, lValues);
+        lValues =
+            getTF()
+                .withSubScope("broadcastWeightsCheck")
+                .withControlDependencies(Collections.singletonList(broadcastWeightsCheck))
+                .math
+                .mul(lValues, lSampleWeights);
+      } catch (IllegalArgumentException ex) {
+        // reduce the values down to the rank of the samples
+        int nDim = lValues.shape().numDimensions();
+        int wDim = lSampleWeights.shape().numDimensions();
+        int numAxes = nDim - wDim;
```

Comment:
Why is `nDim` necessarily `> wDim` at this point?

Reply:
You're right. This is an error, I'll fix it.

Reply:
👍 done